bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
-            timeout: (:type:`int`, optional):
-                Timeout for getting the server's response. 
-        Returns:
-            dict: A dictionary of the files inside of the genesis_datasets and their hashes.
+            max_threads: 
+                The maximum number of threads that can be used to
+                execute the given calls.
         """
-        session = requests.Session()
-        session.params.update((('arg', file_meta['Hash']), ))
-        
-        try:
-            if action == 'get':
-                response = self.requests_retry_session(session=session).get(address, timeout=timeout)
-            elif action == 'post':
-                response = self.requests_retry_session(session=session).post(address, timeout=timeout)
-            logger.success("Loaded from IPFS:".ljust(20) + "<blue>{}</blue>".format(file_meta['Name']))
+        self.max_threads = max_threads
+        self._idle_semaphore = threading.Semaphore(0)
+        self._threads = []
+        self._shutdown_lock = threading.Lock()
+        self._shutdown = False
+
+    def submit(self, fn, args:list=[],kwargs:dict={}):
+        '''
+        Submit a function with args and kwargs on a seperate thread.
+
+        Args
+            fn (Callable):
+                Function to place on the thread.
+            args (list):
+                Arguments to a function.
+            kwargs (dict):
+                Key word arguments to a function.
+        '''
+        with self._shutdown_lock:
+            if self._shutdown:
+                raise RuntimeError('cannot schedule new futures after shutdown')
+            
+            thread = threading.Thread(target=fn, args=args, kwargs=kwargs, daemon=True)
+            thread.start()
+            self._threads.append(thread)
 
-        except Exception as E:
-            logger.error(f"Failed to get from IPFS {file_meta['Name']} {E}")
-            return None
+        return thread
 
-        return response
 
-    def __len__(self):
-        """ Returns length of the dataset that the dataset is processing
-        """
+    @property
+    def threads(self):
+        '''List threads.'''
+        return self._threads
+
+    def __del__(self):
+        self.shutdown()
+
+    def shutdown(self, wait=True):
+        '''Shutdown threads'''
+        if wait:
+            for t in self._threads:
+                t.join()
+
 
-    def __getitem__(self, idx):
-        """ Returns the next batch from the dataset.
-        """
 
-class GenesisTextDataset( Dataset ):
-    """ One kind of dataset that caters for the data from ipfs 
+
+class GenesisTextDataset:
+    """ Implementation for the dataset class, which handles dataloading from ipfs
     """
+
+    ipfs_url = 'http://global.ipfs.opentensor.ai/api/v0'
+    dataset_dir = 'http://global.ipfs.opentensor.ai/api/v0/cat' 
+    text_dir = 'http://global.ipfs.opentensor.ai/api/v0/object/get'
+    mountain_hash = 'QmSdDg6V9dgpdAFtActs75Qfc36qJtm9y8a7yrQ1rHm7ZX'
```

Comment:
These should be defined in a separate file, e.g. `bittensor.__init__` where a lot of other constants are defined.